src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleCloudStorageDAO.scala

```diff
@@ -1,81 +1,226 @@
+    deleteGoogleGroup(WorkspaceAccessLevel.Owner, bucketName, workspaceName)
+  }
+
+  def createGoogleGroupWithACL(bucketCredential: Credential, accessLevel: WorkspaceAccessLevel, bucketName: String, workspaceName: WorkspaceName): Unit = {
+    val groupId = makeGroupId(workspaceName, accessLevel)
+    val group = new Group().setEmail(groupId).setName(s"FireCloud ${workspaceName.namespace}/${workspaceName.name} ${accessLevel}")
+
+    // create group
+    val serviceCredential = getGroupServiceAccountCredential
+    val directory = getGroupDirectory(serviceCredential)
+    directory.groups().insert(group).execute()
+
+    // add group to ACL (requires bucket owner's credential)
+    val storage = getStorage(bucketCredential)
+    val acl = new BucketAccessControl().setEntity(makeGroupEntityString(groupId)).setRole(WorkspaceAccessLevel.toCanonicalString(accessLevel)).setBucket(bucketName)
+    storage.bucketAccessControls().insert(bucketName, acl).execute()
+  }
+
+  def deleteGoogleGroup(accessLevel: WorkspaceAccessLevel, bucketName: String, workspaceName: WorkspaceName): Unit = {
+    val groupId = makeGroupId(workspaceName, accessLevel)
+    val serviceCredential = getGroupServiceAccountCredential
+    val directory = getGroupDirectory(serviceCredential)
+    directory.groups.delete(groupId).execute()
+  }
+
   override def storeUser(userId: String, authCode: String, state: String, callbackPath: String): Unit = {
     if ( userId != state ) throw new RawlsException("Whoa.  We seem to be dealing with the wrong user.")
     val tokenResponse = flow.newTokenRequest(authCode)
-      .setRedirectUri(getOurRedirectURI(callbackPath))
+      .setRedirectUri(getRawlsRedirectURI(callbackPath))
       .execute()
 
     flow.createAndStoreCredential(tokenResponse, userId)
   }
 
-  override def createBucket(userId: String, projectId: String, bucketName: String): Unit = {
-    val credential = getCredential(userId)
-    val storage = new Storage.Builder(httpTransport, jsonFactory, credential).build()
+  override def createBucket(ownerId: String, projectId: String, bucketName: String): Unit = {
+    val credential = getBucketCredential(ownerId)
+    val storage = getStorage(credential)
     val bucket =  new Bucket().setName(bucketName)
+    storage.buckets().insert(projectId, bucket).execute()
+  }
 
-    storage.buckets()
-      .insert(projectId, bucket)
-      .execute()
+  override def getMaximumAccessLevel(userId: String, workspaceName: WorkspaceName): WorkspaceAccessLevel = {
+    val credential = getGroupServiceAccountCredential
+    val directory = getGroupDirectory(credential)
+
+    // note that the iteration order is important here...
+    Seq(WorkspaceAccessLevel.Owner, WorkspaceAccessLevel.Write, WorkspaceAccessLevel.Read) foreach { accessLevel =>
+      try {
+        directory.members().get(makeGroupId(workspaceName, accessLevel), userId).execute()
+        return accessLevel
```

Comment:
Returns are hard to follow in scala and are frowned upon, i.e. I can't tell from looking at the code what this is supposed to return from. Can this be changed into a map instead of foreach with the result being a Seq[Try[WorkspaceAccessLevel]]? If you want to be efficient and not make calls you don't have to, look into using a Stream instead of a Seq.